examples/tinywl-new/wallpaperprovider.cpp

```diff
@@ -0,0 +1,110 @@
+// Copyright (C) 2024 WenHao Peng <pengwenhao@uniontech.com>.
+// SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "wallpaperprovider.h"
+
+#include <QDir>
+#include <QFileInfo>
+#include <QQuickWindow>
+#include <QImageReader>
+#include <QSGTexture>
+
+WallpaperTextureFactory::WallpaperTextureFactory(WallpaperImageProvider* provider, const QImage &image)
+    :m_wallpaperProvider(provider)
+{
+    if (image.format() == QImage::Format_ARGB32_Premultiplied
+        || image.format() == QImage::Format_RGB32) {
+        im = image;
+    } else {
+        im = image.convertToFormat(QImage::Format_ARGB32_Premultiplied);
+    }
+    size = im.size();
+}
+
+WallpaperTextureFactory::WallpaperTextureFactory(WallpaperImageProvider *provider, const QString &id)
+    :m_wallpaperProvider(provider)
+    ,m_wallpaperId(id)
+{
+    m_textureExist = true;
+}
+
+QSGTexture *WallpaperTextureFactory::createTexture(QQuickWindow *window) const
+{
+    if (m_textureExist)
+        m_texture = m_wallpaperProvider->getExistTexture(m_wallpaperId);
+
+    if (!m_texture)
+        m_texture = window->createTextureFromImage(im, QQuickWindow::TextureCanUseAtlas);
+
+    static bool transient = qEnvironmentVariableIsSet("QSG_TRANSIENT_IMAGES");
+    if (transient)
+        const_cast<WallpaperTextureFactory *>(this)->im = QImage();
+    return m_texture;
+}
+
+WallpaperImageProvider::WallpaperImageProvider()
+    : QQuickImageProvider(QQuickImageProvider::Texture)
+{
+}
+
+WallpaperImageProvider::~WallpaperImageProvider()
+{
+    textureCache.clear();
+}
+
+QImage WallpaperImageProvider::loadFile(const QString &path, const QSize &requestedSize)
+{
+    QImageReader imgio(path);
+
+    if (requestedSize.isValid())
+        imgio.setScaledSize(requestedSize);
+
+    QImage image;
+    imgio.read(&image);
+    return image;
+}
+
+QSGTexture *WallpaperImageProvider::getExistTexture(const QString& id) const
+{
+    if (textureCache.contains(id)) {
+        return dynamic_cast<WallpaperTextureFactory*>(textureCache[id])->texture();
+    }
+
+    return nullptr;
+}
+
+QQuickTextureFactory *WallpaperImageProvider::requestTexture(const QString &id, QSize *size, const QSize &requestedSize)
+{
+    QQuickTextureFactory *factory = nullptr;
+    QSize readSize;
+
+    if (textureCache.contains(id)) {
+        auto cache_factory = textureCache[id];
+        readSize = cache_factory->textureSize();
+        if (requestedSize.width() * requestedSize.height() < readSize.width() * readSize.height()) {
```

Comment:
Why is `width * height`?

`if (requestedSize.width() < readSize.width() && requestedSize.height() < readSize.height())`

Reply:
i change